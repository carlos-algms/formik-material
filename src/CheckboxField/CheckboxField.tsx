import { Checkbox, CheckboxProps, FormControl, FormControlLabel } from '@material-ui/core';
import { Field, FieldProps } from 'formik';
import { nanoid } from 'nanoid/non-secure';
import React, { ReactNode } from 'react';
import ErrorLabel from '../ErrorLabel';

export type CheckboxFieldProps = Omit<CheckboxProps, 'form' | 'checked' | 'defaultChecked'> & {
  label?: ReactNode;
};

export default function CheckboxField({
  label = '',
  name = nanoid(),
  className,
  color = 'primary',
  ...checkBoxProps
}: CheckboxFieldProps): JSX.Element {
  return (
    <Field type="checkbox" name={name}>
      {({ field, meta: { error, touched } }: FieldProps) => {
        const { value, checked = false } = field;
        const indeterminate = !Array.isArray(value) && value === null;
        const isErrorState = Boolean(error && touched);

        return (
          <FormControl error={isErrorState} className={className}>
            <FormControlLabel
              label={label}
              control={
                <Checkbox
                  indeterminate={indeterminate}
                  checked={checked}
                  {...checkBoxProps}
                  {...field}
                  color={color}
                />
              }
            />

            {isErrorState && <ErrorLabel error={error} />}
          </FormControl>
        );
      }}
    </Field>
  );
}
