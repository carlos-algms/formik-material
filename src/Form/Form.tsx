import {
  Formik,
  FormikErrors,
  Form as FormikForm,
  FormikHelpers,
  FormikProps,
  FormikValues,
} from 'formik';
import type { ReactNode } from 'react';

export type OnSubmitFormCallback<T = any, R = any> = {
  (values: T, formikHelpers: FormikHelpers<T>): R | Promise<R>;
};

export type FormProps<T extends FormikValues> = {
  // TODO this could extend from FormikConfig<Values> to make it more extendable
  initialValues?: T;
  onSubmitForm: OnSubmitFormCallback<T>;
  validationSchema?: any | (() => any);
  validate?: (values: T) => FormikErrors<T> | Promise<FormikErrors<T>>;
  children?: ReactNode | ((props: FormikProps<T>) => ReactNode);
  className?: string;
};

/**
 * Form Component
 */
export default function Form<T extends FormikValues>({
  children,
  initialValues,
  onSubmitForm,
  validationSchema,
  validate,
  className,
}: FormProps<T>) {
  if (!initialValues) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn('No initial values provided to the <Form> Component');
    }
    return null;
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmitForm}
      validationSchema={validationSchema}
      validate={validate}
    >
      {(formikProps) => {
        return (
          <FormikForm noValidate autoComplete="off" className={className}>
            {typeof children === 'function' ? children(formikProps) : children}
          </FormikForm>
        );
      }}
    </Formik>
  );
}
