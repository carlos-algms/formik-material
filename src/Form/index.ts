export { default } from './Form';
export type { FormProps, OnSubmitFormCallback } from './Form';
