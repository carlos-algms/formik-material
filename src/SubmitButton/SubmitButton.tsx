import Button from '@material-ui/core/Button';
import { useFormikContext } from 'formik';
import React, { memo } from 'react';
import type PropsBase from '../shared/PropsBase';

function SubmitButton({ children, className }: PropsBase) {
	const { isSubmitting } = useFormikContext();

	return (
		<div className={className}>
			<Button variant="contained" color="primary" type="submit" disabled={isSubmitting}>
				{children}
			</Button>
		</div>
	);
}

export default memo(SubmitButton);
