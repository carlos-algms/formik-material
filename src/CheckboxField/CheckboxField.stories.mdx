import { action } from '@storybook/addon-actions';
import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs/blocks';
import SubmitButton from '../SubmitButton';
import useFormSubmittedValues from '../useFormSubmittedValues';
import CheckboxField from './CheckboxField';
import FormDemo from '../shared/FormDemo';
import Typography from '@mui/material/Typography';
import * as yup from 'yup';

export const initialValues = {
  read: false,
  accept: true,
};

<Meta title="Checkbox Field" component={CheckboxField} />

<Typography variant="h4">Checkbox Field</Typography>

<Canvas>
  <Story name="basic">
    <FormDemo onSubmit={action('submitted')} initialValues={initialValues}>
      <Typography variant="h5">Basic Example</Typography>
      <CheckboxField name="read" label="Read and accept our terms" />
      <CheckboxField name="accept" label="I accept receiving newsletter" color="secondary" />
    </FormDemo>
  </Story>
</Canvas>

export const validationSchema = yup.object().shape({
  read: yup.boolean().required().oneOf([true]),
  accept: yup.boolean().required(),
});

<Canvas>
  <Story name="validation">
    <FormDemo
      onSubmit={action('submitted valid')}
      initialValues={initialValues}
      validationSchema={validationSchema}
    >
      <Typography variant="h5">Validation</Typography>
      <CheckboxField name="read" label="Read and accept our terms" />
      <CheckboxField name="accept" label="I accept receiving newsletter" color="secondary" />
    </FormDemo>
  </Story>
</Canvas>

## Props

<ArgsTable of={CheckboxField} />
