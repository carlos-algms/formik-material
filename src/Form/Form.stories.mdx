import { action } from '@storybook/addon-actions';
import { Meta, Canvas, Story, Props } from '@storybook/addon-docs/blocks';
import Form from './Form';
import SubmitButton from '../SubmitButton';
import useFormSubmittedValues from '../useFormSubmittedValues';
import TextField from '../TextField';

export const initialValues = { name: 'initial name' };

<Meta title="Form Component" component={Form} />

## Basic form implementation

<Canvas>
  <Story name="basic">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          <TextField name="name" label="Name: " />
          <p>
            <SubmitButton>Submit</SubmitButton>
          </p>
        </Form>
      );
    }}
  </Story>
</Canvas>

## Forms should be able to render child functions

<Canvas>
  <Story name="render child function">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          {() => (
            <>
              <TextField name="name" label="Name: " />
              <p>
                <SubmitButton>Submit</SubmitButton>
              </p>
            </>
          )}
        </Form>
      );
    }}
  </Story>
</Canvas>
