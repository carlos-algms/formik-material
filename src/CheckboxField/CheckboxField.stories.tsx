import Typography from '@mui/material/Typography';
import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';
import * as yup from 'yup';
import FormDemo from '../shared/FormDemo';
import CheckboxField from './CheckboxField';

const initialValues = {
  read: false,
  accept: true,
};

const meta: Meta<typeof CheckboxField> = {
  title: 'Checkbox Field',
  component: CheckboxField,
};

export default meta;

type Story = StoryObj<typeof CheckboxField>;

export const Basic: Story = {
  render: () => (
    <FormDemo onSubmit={action('submitted')} initialValues={initialValues}>
      <Typography variant="h5">Basic Example</Typography>
      <CheckboxField name="read" label="Read and accept our terms" />
      <CheckboxField name="accept" label="I accept receiving newsletter" color="secondary" />
    </FormDemo>
  ),
};

export const Validation: Story = {
  render: () => (
    <FormDemo
      onSubmit={action('submitted valid')}
      initialValues={initialValues}
      validationSchema={validationSchema}
    >
      <Typography variant="h5">Validation</Typography>
      <CheckboxField name="read" label="Read and accept our terms" />
      <CheckboxField name="accept" label="I accept receiving newsletter" color="secondary" />
    </FormDemo>
  ),
};

const validationSchema = yup.object().shape({
  read: yup.boolean().required().oneOf([true]),
  accept: yup.boolean().required(),
});
