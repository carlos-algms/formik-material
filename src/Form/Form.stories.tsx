import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';
import SubmitButton from '../SubmitButton';
import TextField from '../TextField';
import useFormSubmittedValues from '../useFormSubmittedValues';
import Form from './Form';

const initialValues = { name: 'initial name' };

const meta: Meta<typeof Form> = {
  title: 'Form Component',
  component: Form,
};

export default meta;

type Story = StoryObj<typeof Form>;

export const Basic: Story = {
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        <TextField name="name" label="Name: " />
        <p>
          <SubmitButton>Submit</SubmitButton>
        </p>
      </Form>
    );
  },
};

export const ChildFn: Story = {
  name: 'Render children as function',
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        {() => (
          <>
            <TextField name="name" label="Name: " />
            <p>
              <SubmitButton>Submit</SubmitButton>
            </p>
          </>
        )}
      </Form>
    );
  },
};
