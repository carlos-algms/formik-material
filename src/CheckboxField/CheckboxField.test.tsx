import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import Form from '../Form';
import SubmitButton from '../SubmitButton';
import { delay } from '../tests/helpers';
import CheckboxField from './CheckboxField';

test('should render a MUI Checkbox and error label', async () => {
  const submitMock = jest.fn();
  const values = { testField: false };
  const validateMock = jest.fn(() => ({
    testField: 'Not valid',
  }));

  const { container, getByRole } = render(
    <Form onSubmitForm={submitMock} initialValues={values} validate={validateMock}>
      <CheckboxField name="testField" label="Test Checkbox" className="checkbox-test-class" />
      <SubmitButton>Send</SubmitButton>
    </Form>,
  );

  const submitButton = getByRole('button');

  await act(async () => {
    fireEvent.click(submitButton);
    await delay();
  });

  const wrapper = container.querySelector('.checkbox-test-class');
  expect(wrapper).toMatchSnapshot();
});
