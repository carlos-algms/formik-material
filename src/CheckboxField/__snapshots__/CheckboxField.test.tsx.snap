// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a MUI Checkbox and error label 1`] = `
<div
  class="MuiFormControl-root checkbox-test-class css-1nrlq1o-MuiFormControl-root"
>
  <label
    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd Mui-error css-j204z7-MuiFormControlLabel-root"
  >
    <span
      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-1jg3gx8-MuiButtonBase-root-MuiCheckbox-root"
    >
      <input
        class="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate="false"
        name="testField"
        type="checkbox"
        value="false"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
    >
      Test Checkbox
    </span>
  </label>
  <div
    class="MuiTypography-root MuiTypography-body2 error-label css-1lt5m6e-MuiTypography-root"
  >
    Not valid
  </div>
</div>
`;
